### PR TITLE
feature(events): Adds new handler API available for events and hooks

### PR DESCRIPTION
**Pull request by mrclay**
Handlers can type-hint `Elgg\Hook` or `Elgg\Event` to receive a single object with various methods.

Class names can be registered as handlers. Each time the hook/event is triggered, the class is instantiated and invoked.

### CODE REVIEW

File: engine/classes/Elgg/HooksRegistrationService/Hook.php [hunk 1]
```diff
@@ -0,0 +1,120 @@
...
+		if (!is_array($this->params)) {
+			return $default;
+		}
+		return array_key_exists($key, $this->params) ? $this->params[$key] : $default;
+	}
+
+	/**
+	 * {@inheritdoc}
+	 */
+	public function getEntity() {
```
Comment (mrclay) [inline engine/classes/Elgg/HooksRegistrationService/Hook.php]: getEntityParam() and getUserParam()

**Comment by mrclay** [comment]
@hypeJunction This now has all the features of #8399, but auto-detects whether to pass in the single hook object or the traditional args.

File: docs/design/events.rst [hunk 2]
```diff
@@ -110,24 +128,49 @@ in the framework: system, user, object, relationship, annotation, group.
...
+    namespace MyPlugin;
+
+    class UpdateObjectHandler {
+        public function __invoke(\Elgg\ObjectEvent $event) {
+
+        }
+    }
+
+    // in init, system
+    elgg_register_event_handler('update', 'object', UpdateObjectHandler::class);
```
Comment (hypeJunction) [inline docs/design/events.rst]: Namespace?

File: docs/design/events.rst [hunk 3]
```diff
@@ -175,9 +220,20 @@ trigger.
...
+.. code-block:: php
+
+    <?php
+
+    /**
+     * @param \Elgg\Hook $hook The hook object
+     *
+     * @return mixed if not null, this will be the new value of the plugin hook
+     */
+    function plugin_hook_handler1(\Elgg\Hook $hook, $type, $value, $params) {
```
Comment (hypeJunction) [inline docs/design/events.rst]: Do we still need all the args?

**Comment by hypeJunction** [comment]
Very positive first impression.  I will dig deeply into it later.
I think the docs will confuse newcomers with several ways of doing things. I think we should document the preffered way in more detail and add a legacy section.

**Comment by mrclay** [comment]
On my system this adds about 20ms to the logged out /activity page `ab -t 10 </activity>`.

**Comment by mrclay** [comment]
Creating event/hook objects on first demand brings the impact down to 5ms (3%). I can't see any more ways to reduce the cost, but I'll keep playing with it.

**Comment by mrclay** [comment]
Latest comparison with 2.x adds only 1ms.

**Comment by hypeJunction** [comment]
Cool. I will dig into it next week. Spent too much time on Elgg this week.

**Comment by hypeJunction** [comment]
I think this should wait till 3.0

**Comment by mrclay** [comment]
Rebased to master. ab-ing a discussion page logged out, this brings median time from 98ms to 105ms.

**Comment by hypeJunction** [comment]
Looks great. I am happy to sacrifice performance for dev usability.

**Review by hypeJunction — APPROVED** [review]

**Comment by hypeJunction** [comment]
Plus, this will be a great foundation to move actions to classes.

**Comment by hypeJunction** [comment]
I am pulling the trigger on this.

**Pull request merged**